core/src/main/kotlin/fr/sncf/osrd/sim_infra_adapter/RawInfraFromRjsAdapter.kt

```diff
@@ -278,23 +276,113 @@ val allLoadingGaugeTypeSet = enumValues<RJSLoadingGaugeType>().toSet()
 private fun parseRjsRouteWaypoint(
     rjsWaypoint: RJSRouteWaypoint,
     builder: RawInfraFromRjsBuilderImpl,
-    trackSectionNameToChunkEndOffset: MutableMap<String, TreeSet<Offset<TrackSection>>>,
+    trackSectionNameToDistanceSortedWaypointIdxs:
+        MutableMap<String, TreeMap<Offset<TrackSection>, DetectorId>>,
     detectorMap: MutableMap<String, DetectorId>,
 ) {
-    detectorMap[rjsWaypoint.id] = builder.detector(rjsWaypoint.id)
+    val waypointIdx = builder.detector(rjsWaypoint.id)
+    detectorMap[rjsWaypoint.id] = waypointIdx
 
-    // Storing only chunk's end: ignore 0 offset that is a chunk's start
-    if (rjsWaypoint.position == 0.0) return
+    if (!trackSectionNameToDistanceSortedWaypointIdxs.containsKey(rjsWaypoint.track)) {
+        trackSectionNameToDistanceSortedWaypointIdxs[rjsWaypoint.track] =
+            TreeMap<Offset<TrackSection>, DetectorId>()
+    }
+    trackSectionNameToDistanceSortedWaypointIdxs[rjsWaypoint.track]!![
+        Offset(rjsWaypoint.position.meters)] = waypointIdx
+}
 
-    if (!trackSectionNameToChunkEndOffset.containsKey(rjsWaypoint.track)) {
-        trackSectionNameToChunkEndOffset[rjsWaypoint.track] = TreeSet<Offset<TrackSection>>()
+/** Build detection zones on a partially-filled builder (track-sections, nodes and detectors) */
+private fun buildZones(
+    builder: RawInfraFromRjsBuilderImpl,
+    trackSectionNameToDistanceSortedWaypointIdxs:
+        Map<String, TreeMap<Offset<TrackSection>, DetectorId>>
+) {
+    // run a union-find to associate track endpoints:
+    // union by node
+    val trackSectionPool = builder.getTrackSectionPool()
+    val trackNodePool = builder.getTrackNodePool()
+    val uf = UnionFind(trackSectionPool.size.toInt() * 2)
+    for (nodeIdx in trackNodePool) {
+        val ports = trackNodePool[nodeIdx].ports
+        var firstEndpoint: EndpointTrackSectionId? = null
+        for (port in ports) {
+            val endpoint = ports[port]
+            if (firstEndpoint == null) {
+                firstEndpoint = endpoint
+            } else {
+                uf.union(firstEndpoint.index.toInt(), endpoint.index.toInt())
+            }
+        }
+    }
+    // union by track-section without waypoint
+    for (trackSectionIdx in trackSectionPool) {
+        val waypoints =
+            trackSectionNameToDistanceSortedWaypointIdxs[trackSectionPool[trackSectionIdx].name]
```

Comment:
waypoint is the legacy name for detectors. we try not to use it anymore, as it also refers to pathfinding waypoints

Reply:
OK, I kept `waypoint` to wrap detectors and buffer-stops.
Considering that, if you confirm that `detector` is better, I'm OK for the change (I don't have a 3rd alternative in mind).

Reply:
Done in https://github.com/osrd-project/osrd/pull/6921/commits/0f93e12e301ca1b649b1b10a40fcd271961674f6